### ISSUE TITLE
Provide a replacement for the previous CABPK reference implementation for implementors

### ISSUE THREAD

**Issue by detiber** [kubernetes-sigs/cluster-api#1559]
**What steps did you take and what happened:**

With #1545 we will no longer be able to point folks interested in creating a bootstrap provider towards CABPK as a reference implementation. We need to provide a replacement for this (could be docs, could be a test reference implementation, etc). The replacement should be sufficient to get started with a bootstrap provider without having to disentangle what is CAPI related vs CABPK related in the cluster-api repo.

/kind documentation
/priority important-soon
/milestone v0.3.0

related to: #1090 #1138

**Comment by avorima** [comment]
I'd like to tackle this. My plan is to add a reference implementation under `bootstrap/example`.

**Comment by detiber** [comment]
@avorima while I think it might be possible to do a full reference implementation as a submodule, I think it might be a bit too easy to accidentally re-use some of the top level tooling. It might be better to take a documentation first approach like this PR is doing for infrastructure providers: https://github.com/kubernetes-sigs/cluster-api/pull/1682

**Comment by avorima** [comment]
Good point. I'm using the CAPI `docs/**/*.md` files for a lot of reference as well. I'll write some docs then.

**Comment by ncdc** [comment]
@avorima are you still interested in working on this?

**Comment by avorima** [comment]
@ncdc Sorry for the late response. Yes, I'd still like to finish this.

Regarding the docs: IMO the initial steps for creating a bootstrap provider are implicitly documented in the implementer's guide for cluster providers. Should I assume the reader is aware of this and focus on the steps that are specific to bootstrap providers or should it be a self-contained section?

**Comment by ncdc** [comment]
@avorima I'm tempted to say self-contained, or find some way to reorganize the cluster one so that there's a common section and then sections specific to cluster & bootstrap providers. @vincepri wdyt?

**Comment by vincepri** [comment]
This should be potentially an extension of the current guide we already have today https://cluster-api.sigs.k8s.io/developer/providers/implementers-guide/overview.html
